src/object/obj_tx.c

```diff
@@ -0,0 +1,958 @@
+		D_ERROR("TX is READ ONLY\n");
+		D_GOTO(out, rc = -DER_NO_PERM);
+	}
+
+out:
+	if (rc != 0)
+		dc_tx_decref(tx);
+	else
+		*ptx = tx;
+
+	return rc;
+}
+
+static void
+dc_tx_cleanup(struct dc_tx *tx)
+{
+	struct dc_tx_sub_req	*dtsr;
+	int			 i;
+
+	while ((dtsr = d_list_pop_entry(&tx->tx_sub_reqs,
+					struct dc_tx_sub_req,
+					dtsr_link)) != NULL) {
+		switch (dtsr->dtsr_opc) {
+		case DAOS_OBJ_RPC_UPDATE:
+			if (dtsr->dtsr_zero_copy)
+				break;
+
+			for (i = 0; i < dtsr->dtsr_nr; i++) {
+				daos_iov_free(&dtsr->dtsr_iods[i].iod_name);
+				D_FREE(dtsr->dtsr_iods[i].iod_recxs);
+				daos_sgl_fini(&dtsr->dtsr_sgls[i], true);
+			}
+
+			D_FREE(dtsr->dtsr_iods);
+			D_FREE(dtsr->dtsr_sgls);
+			daos_iov_free(dtsr->dtsr_dkey);
+			break;
+		case DAOS_OBJ_RPC_PUNCH:
+			break;
+		case DAOS_OBJ_RPC_PUNCH_DKEYS:
+			if (!dtsr->dtsr_zero_copy)
+				daos_iov_free(dtsr->dtsr_dkey);
+			break;
+		case DAOS_OBJ_RPC_PUNCH_AKEYS:
+			if (!dtsr->dtsr_zero_copy) {
+				for (i = 0; i < dtsr->dtsr_nr; i++)
+					daos_iov_free(&dtsr->dtsr_akeys[i]);
+
+				D_FREE(dtsr->dtsr_akeys);
+				daos_iov_free(dtsr->dtsr_dkey);
+			}
+			break;
+		default:
+			D_ASSERT(0);
+		}
+
+		obj_decref(dtsr->dtsr_obj);
+		D_FREE(dtsr);
+	}
+}
+
+int
+dc_tx_hdl2epoch_with_check(daos_handle_t th, bool check_write,
+			   daos_epoch_t *epoch, uint32_t *pm_ver)
+{
+	struct dc_tx	*tx = NULL;
+	daos_handle_t	 ph;
+	int		 rc;
+
+	rc = dc_tx_check(th, check_write, &tx);
+	if (rc != 0)
+		return rc;
+
+	ph = dc_cont_hdl2pool_hdl(tx->tx_coh);
+	D_ASSERT(daos_handle_is_inval(ph));
+
+	rc = dc_pool_map_version_get(ph, pm_ver);
+	if (rc != 0)
+		D_GOTO(out, rc);
+
+	if (tx->tx_pm_ver == 0) {
+		tx->tx_pm_ver = *pm_ver;
+	} else if (tx->tx_pm_ver != *pm_ver) {
+		D_ASSERTF(tx->tx_pm_ver < *pm_ver,
+			  "Pool map version is reverted from %u to %u\n",
+			  tx->tx_pm_ver, *pm_ver);
+
+		tx->tx_pm_ver = *pm_ver;
+
+		/* If the pool map is refreshed, then it is possible that
+		 * the data from former fetch/list/query become stale. On
+		 * the other hand, even if related data is still valid,
+		 * but related read timestamp (used for MVCC) may be left
+		 * on the server that has been evicted from the cluster.
+		 * So we have to restart the transaction under such case.
+		 */
+		D_GOTO(out, rc = -DER_AGAIN);
+	}
+
+	*epoch = daos_dti2epoch(&tx->tx_id);
+
+out:
+	dc_tx_decref(tx);
+
+	return rc;
+}
+
+int
+dc_tx_hdl2epoch(daos_handle_t th, daos_epoch_t *epoch)
+{
+	struct dc_tx	*tx;
+
+	if (daos_handle_is_inval(th))
+		return -DER_INVAL;
+
+	tx = dc_tx_hdl2ptr(th);
+	if (tx == NULL)
+		return -DER_NO_HDL;
+
+	*epoch = daos_dti2epoch(&tx->tx_id);
+	dc_tx_decref(tx);
+
+	return 0;
+}
+
+int
+dc_tx_open(tse_task_t *task)
+{
+	daos_tx_open_t	*args;
+	struct dc_tx	*tx = NULL;
+	int		 rc;
+
+	args = dc_task_get_args(task);
+	D_ASSERTF(args != NULL,
+		  "Task Argument OPC does not match DC OPC (open)\n");
+
+	rc = dc_tx_alloc(args->coh, crt_hlc_get(), 0, &tx);
+	if (rc == 0);
```

Comment:
(style)  trailing semicolon indicates no statements, indent implies otherwise
(style)  trailing statements should be on next line